### PR TITLE
Add File Globbing Support for Ignore Files/Folders

### PR DESCRIPTION
**Pull request by kaizen365**
This PR adds file globbing support for Ignoring Files/Folders as discussed in https://github.com/excubo-ag/WebCompiler/pull/58#issue-1099746053.

Note: This PR should be either reviewed/merged after #58 merges (as was based off the same branch and includes _Support Trailing Slash in IgnoreFolder variables_ commit or #58 abandoned in favour of this PR which has both commits.

```
@kaizen365 ➜ /workspaces/WebCompiler (feature/ignorefolder-globbing ✗) $ dotnet test
  Determining projects to restore...
  All projects are up-to-date for restore.
  WebCompiler -> /workspaces/WebCompiler/WebCompiler/bin/Debug/net6.0/Excubo.WebCompiler.dll
  Tests_WebCompiler -> /workspaces/WebCompiler/Tests_WebCompiler/bin/Debug/net6.0/Tests_WebCompiler.dll
Test run for /workspaces/WebCompiler/Tests_WebCompiler/bin/Debug/net6.0/Tests_WebCompiler.dll (.NETCoreApp,Version=v6.0)
Microsoft (R) Test Execution Command Line Tool Version 17.0.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:    50, Skipped:     0, Total:    50, Duration: 13 s - /workspaces/WebCompiler/Tests_WebCompiler/bin/Debug/net6.0/Tests_WebCompiler.dll (net6.0)
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -23,14 +23,25 @@ As a benefit, this implementation is cross-platform (x64 linux/win are tested, p
...
+```json
+  "CompilerSettings": {
+     "Ignore": [ "wwwroot/*.scss", "wwwroot/css/specific-file.scss", "wwwroot/_lib/**/*.scss", "bin/**/*", "obj/**/*" ],
+  }
+```
+
 #### Changes in version 3.2.Y
 
-Supports excluding certain files and folders when using recursive mode
+Supports excluding certain files and folders (nb. Sub folders must be specified specifically if wanting to be ignored) when using recursive mode
```
Comment (stefanloerwald) [inline README.md:39]: This wouldn't be the case with v3.3.0, right? My expectation for ignoring a folder is that the tool never visits that folder, i.e. including all subfolders.

File: Tests_WebCompiler/WholeProgramTests.cs [hunk 2]
```diff
@@ -453,15 +453,21 @@ public void ScssFilesAndFoldersConfiguredToBeIgnoredAreIgnored()
...
             {
-                // suppressed by IgnoreFolders
+                // suppressed by Ignore
                 "../../../TestCases/Scss/IgnoreFolder/globalVariables.css",
+                "../../../TestCases/Scss/IgnoreFolder2/globalVariables.css",
+                "../../../TestCases/Scss/IgnoreFolder2/SubFolder/test.css",
+                "../../../TestCases/Scss/IgnoreFolderAndSubFolders/globalVariables.css",
+                "../../../TestCases/Scss/IgnoreFolderAndSubFolders/SubFolder1/test.css",
+                "../../../TestCases/Scss/IgnoreFolderAndSubFolders/SubFolder2/test.css",
                 // suppressed by IgnoreFiles
```
Comment (stefanloerwald) [inline Tests_WebCompiler/WholeProgramTests.cs:471]: IgnoreFiles does not exist anymore: remove

File: WebCompiler/Helpers/FileFolderHelpers.cs [hunk 3]
```diff
@@ -6,21 +7,16 @@ namespace WebCompiler.Helpers
...
         {
-            foreach (var file in Directory.GetFiles(directory))
-            {
-                if (!Path.GetFileName(file).StartsWith('_'))
-                {
-                    yield return file;
-                }
-            }
-            foreach (var subdir in Directory.GetDirectories(directory))
+            ignoreGlobs.Add("**/_*.*"); // Also Ignore all files that start with _
```
Comment (stefanloerwald) [inline WebCompiler/Helpers/FileFolderHelpers.cs:12]: Related bug: https://github.com/excubo-ag/WebCompiler/issues/54

If you see an easy way to support this as well, please consider!

File: WebCompiler/WebCompiler.csproj [hunk 4]
```diff
@@ -20,6 +20,7 @@
     <PackageReference Include="Microsoft.ClearScript.V8.Native.linux-arm64" Version="7.2.0" />
     <PackageReference Include="Microsoft.ClearScript.V8.Native.osx-x64" Version="7.2.0" />
     <PackageReference Include="Microsoft.ClearScript.V8.Native.osx-arm64" Version="7.2.0" />
+    <PackageReference Include="Microsoft.Extensions.FileSystemGlobbing" Version="6.0.0" />
```
Comment (stefanloerwald) [inline WebCompiler/WebCompiler.csproj:23]: What a neat package, didn't even know about it! Thanks for using it and not reinventing the wheel (as I would have probably tried).

File: WebCompiler/Helpers/FileFolderHelpers.cs (hunk 3 above)
Comment (kaizen365) [reply WebCompiler/Helpers/FileFolderHelpers.cs:12]: This is a good question. A breaking change could be that we remove this explicit rule in the code and suggest a default template in the README file with this explicitly noted?

**Example**
```json
  "CompilerSettings": {
     "Ignore": [ "bin/**/*", "obj/**/*", "wwwroot/*.scss", "wwwroot/css/specific-file.scss", "wwwroot/_lib/**/*.scss", "**/_*.*" ],
  }
```

File: WebCompiler/WebCompiler.csproj (hunk 4 above)
Comment (kaizen365) [reply WebCompiler/WebCompiler.csproj:23]: Yes, it's quite neat isn't it. I wasn't personally aware of it until it was pointed out by one of our team members (@HennyH) here on a project we're using webcompiler on.

**Off-Topic**
I was trialling out excubo.webcompiler for our sass compilation in conjunction with buildbundleminifier for ourminify/bundle workflow for a few of projects currently working on but am currently favouring switched back to npm and gulp tasks as performance/integration-workflow appears to be better - in particular for hot-reload work.

Will revisit this as time allows and see what is missing in excubo.webcompiler (and buildbundleminifier - as the deficiencies might not be in your webcompiler packaging) compared to the npm + gulp tasks as I really like what you have done here with the .NET implementation.

Mentioning this as I note that in one of the issues you mention that you aren't using your own library in your other projects - curious to see what other limitations you may have found and whether we should consider having an open issue(s) listing these limitations for other contributors potentially to resolve?

File: README.md (hunk 1 above)
Comment (kaizen365) [reply README.md:39]: Yes v3.3 replaces the distinction between IgnoreFiles and IgnoreFolders in favour of an Ignore key with File Globbing.
I've left these notes in 3.2 as 3.2 still exists (albeit being a preview version we could remove these notes maybe?)

Comment (stefanloerwald) [reply README.md:39]: I think the notes should be kept. Doesn't hurt in any case.

File: WebCompiler/Helpers/FileFolderHelpers.cs (hunk 3 above)
Comment (stefanloerwald) [reply WebCompiler/Helpers/FileFolderHelpers.cs:12]: What happens if we set the default `Ignore` value to `new List<string> { ""**/_*.*" }`? Shouldn't that keep the behavior as-is and make it easier to opt out?

File: WebCompiler/WebCompiler.csproj (hunk 4 above)
Comment (stefanloerwald) [reply WebCompiler/WebCompiler.csproj:23]: My use cases didn't require frequent re-compilation of scss, in fact, I barely used it. I used it for one bigger thing (material design, one-time compilation of scss for the basic theme), and used only css for the rest.

The choice of right tooling is very much individual/project-based. As long as a tool is useful to you and doesn't cause you many headaches, just use it until an even better tool comes along ;-)

File: WebCompiler/Helpers/FileFolderHelpers.cs (hunk 3 above)
Comment (kaizen365) [reply WebCompiler/Helpers/FileFolderHelpers.cs]: Done 👍

**Comment by stefanloerwald** [comment]
Thanks @kaizen365! I've just merged and deployed a new version.

**Pull request merged**